### PR TITLE
[server] Cleanup around EntitlementService and BillingMode

### PR DESCRIPTION
**Pull request by geropl**
## Description

Blocked by: https://github.com/gitpod-io/gitpod/pull/18412

This PR refactors the EntitlementService interface, which we did not follow up on after our move to Stripe, but is in the way of some refactorings around GitpodServerImpl and WorkspaceStarter.

This does container a "breaking change", though: We take away some features in non-paid orgs ([details](https://github.com/gitpod-io/gitpod/pull/18429#discussion_r1283498750))

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at bd7191c</samp>

This pull request refactors and simplifies some of the code related to billing and entitlement logic in the server component. It removes unused or redundant parameters, imports, and code, and uses more consistent and concise method calls. It affects the files `billing-mode.ts`, `entitlement-service-ubp.ts`, `entitlement-service.ts`, `gitpod-server-impl.ts`, `workspace-starter.ts`, and `prebuild-manager.ts`.

</details>

## Related Issue(s)

Fixes #

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/billing/billing-mode.ts [hunk 1]
```diff
@@ -33,7 +32,7 @@ export class BillingModes {
         return { mode: "usage-based", paid };
     }
 
-    async getBillingModeForUser(user: User, now: Date): Promise<BillingMode> {
+    async getBillingModeForUser(): Promise<BillingMode> {
```
Comment (svenefftinge) [inline components/server/src/billing/billing-mode.ts:35]: `getBillingModeForUser` without any user context looks pretty odd from the outside. Do we need this method at all?

File: components/server/src/billing/entitlement-service-ubp.ts [hunk 2]
```diff
@@ -63,69 +62,63 @@ export class EntitlementServiceUBP implements EntitlementService {
...
-    async limitNetworkConnections(user: User, date: Date): Promise<boolean> {
+    async limitNetworkConnections(userId: string, organizationId: string): Promise<boolean> {
         // gpl: Because with the current payment handling (pay-after-use) having a "paid" plan is not a good enough classifier for trushworthyness atm.
         // We're looking into improving this, but for the meantime we limit network connections for everybody to reduce the impact of abuse.
         return true;
     }
 
-    private async hasPaidSubscription(user: User, date: Date): Promise<boolean> {
+    private async hasPaidSubscription(userId: string, organizationId?: string): Promise<boolean> {
+        if (organizationId) {
```
Comment (svenefftinge) [inline components/server/src/billing/entitlement-service-ubp.ts:112]: This is good and we should even aim for making the organization mandatory, but it is a breaking change. For instance, users who were able to change timeouts on workspaces running on their free org (but have another paid org) can now no longer do this.

**Review by svenefftinge — APPROVED** [review]
I like this change and believe we can just roll it out. I would even consider it a bug that I would have all these paid features on a non-paid org before. So thanks for fixing! :-)

File: components/server/src/billing/entitlement-service-ubp.ts (hunk 2 above)
Comment (geropl) [reply components/server/src/billing/entitlement-service-ubp.ts:112]: Exactly. But sharpening that behavior is sth we should have done all along, right after the migration to Stripe.

File: components/server/src/billing/billing-mode.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/billing/billing-mode.ts:39]: I added a comment, and explicitly marked it as deprecated now. All we really need to do is add orgnaizationId to `api.maySetTimeout` and `api.updateWorkspaceTimeoutSetting`, then we can remove it entirely.

**Pull request merged**